pkg/collect/utilisation.go

```diff
@@ -0,0 +1,363 @@
+package collect
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"strconv"
+	"time"
+
+	"github.com/mesosphere/dkp-cli-runtime/core/output"
+	coreout "github.com/mesosphere/dkp-cli-runtime/core/output"
+	"github.com/phayes/freeport"
+	"github.com/pkg/errors"
+	prom "github.com/prometheus/client_golang/api"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+	troubleshootv1beta2 "github.com/replicatedhq/troubleshoot/pkg/apis/troubleshoot/v1beta2"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/httpstream"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/client-go/transport/spdy"
+)
+
+type UsageReading struct {
+	Timestamp time.Time `json:"timestamp,omitempty" yaml:"timestamp,omitempty"`
+	CPU       float64   `json:"cpu,omitempty" yaml:"cpu,omitempty"`
+	Memory    float64   `json:"memory,omitempty" yaml:"memory,omitempty"`
+}
+
+type Utilisation struct {
+	Readings map[string][]UsageReading `json:"readings,omitempty" yaml:"readings,omitempty"`
+}
+
+type CollectCPUMemUtilisation struct {
+	Collector    *troubleshootv1beta2.CPUMemUsage
+	BundlePath   string
+	Namespace    string
+	ClientConfig *rest.Config
+	Client       kubernetes.Interface
+	Context      context.Context
+	RBACErrors
+}
+
+func (c *CollectCPUMemUtilisation) Title() string {
+	return getCollectorName(c)
+}
+
+func (c *CollectCPUMemUtilisation) IsExcluded() (bool, error) {
+	return isExcluded(c.Collector.Exclude)
+}
+
+// todo: Rework.
+func (c *CollectCPUMemUtilisation) Collect(progressChan chan<- interface{}) (CollectorResult, error) {
+	thanosServiceNamespace := "kommander"
+	output := NewResult()
+
+	// Get the pods of the deployment
+	pods, err := c.Client.CoreV1().Pods(thanosServiceNamespace).List(context.Background(), metav1.ListOptions{
+		LabelSelector: "app.kubernetes.io/component=query,app.kubernetes.io/instance=thanos,app.kubernetes.io/name=thanos",
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "Could not look up thanos-query pods: ")
+	}
+	if len(pods.Items) < 1 {
+		// Check if no thanos-query pods are found.
+		return nil, errors.New("Could not find any thanos-query pods.")
+	}
+	podName := pods.Items[0].Name
+
+	// Create a Prometheus API client using the API server endpoint
+	prometheusClient, stopCh, err := StartPortForwarding(context.Background(), c.ClientConfig, coreout.NewNonInteractiveShell(os.Stdout, os.Stderr, 4), podName)
+	if err != nil {
+		return nil, errors.Wrap(err, "Error querying Prometheus for CPU utilization:")
+	}
+	v1api := v1.NewAPI(prometheusClient)
+
+	// Query for CPU utilization
+	cpuQuery := "sum(rate(container_cpu_usage_seconds_total{container!='POD', container!=''}[1h])) by (namespace)"
+	start := time.Now().Add(-1 * time.Duration(c.Collector.Duration) * 24 * time.Hour)
```

Comment:
what is `time.Duration(c.Collector.Duration) * 24` ?

Reply:
https://github.com/mesosphere/troubleshoot/blob/36526358087ae02f5de0df80644b05439a86a1f7/pkg/apis/troubleshoot/v1beta2/collector_shared.go#L32